js/eventloop.go

```diff
@@ -0,0 +1,165 @@
+package js
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/dop251/goja"
+)
+
+// A basic event loop implemnentation, with a cut down setTimeout and
+// handling of unhandled rejected promises
+//
+// A specific thing about this event loop is that it will wait to return
+// not only until the queue is empty but until nothing is reserved.
+// This is in contrast with more common behaviours where it only returns on
+// a specific event/action or when the loop is empty.
+// This is required as in k6 iterations (for which event loop will be primary used)
+// are supposed to be independent and any work started in them needs to finish,
+// but also they need to end when all the instructions are done.
+// Additionally because of this on any error while the event loop will exit it's
+// required to wait on the event loop to be empty before the execution can  continue.
+type eventLoop struct {
+	lock          sync.Mutex
+	queue         []func() error
+	wakeupCh      chan struct{} // maybe use sync.Cond ?
+	reservedCount int
+
+	// pendingPromiseRejections are rejected promises with no handler,
+	// if there is something in this map at an end of an event loop then it will exit with an error.
+	// It's similar to what Deno and Node do.
+	pendingPromiseRejections map[*goja.Promise]struct{}
+}
+
+// newEventLoop returns a new event loop with a few helpers attach on it:
+// - adding setTimeout javascript implementation
+// - reporting (and aborting on) unhandled promise rejections
+func newEventLoop(ctx context.Context, rt *goja.Runtime) *eventLoop {
+	e := &eventLoop{
+		wakeupCh:                 make(chan struct{}, 1),
+		pendingPromiseRejections: make(map[*goja.Promise]struct{}),
+	}
+	rt.SetPromiseRejectionTracker(e.promiseRejectionTracker)
+	e.addSetTimeout(ctx, rt)
+
+	return e
+}
+
+func (e *eventLoop) wakeup() {
+	select {
+	case e.wakeupCh <- struct{}{}:
+	default:
+	}
+}
+
+// reserve "reserves" a spot on the loop, preventing it from returning/finishing. The returning function will queue it's
+// argument and wakeup the loop if needed.
+// If the eventLoop has since stopped it will not be execeted
+// This function *must* be called from within running on the event loop, but it's result can be called from anywhere
+// TODO better name
```

Comment:
Any suggestions ? ;)

Reply:
My bad, I didn't reiterate my suggestion because we already had [that discussion](https://github.com/grafana/k6/pull/2228#discussion_r769745602) 😄 I don't have a better idea than what was said previously, but I do think that we should either rename it, or remove the `TODO` anyway. @codebien, any ideas to contribute on that front maybe? 😉

Reply:
`registerCallback` I guess is slightly better than `Reserve` :thinking: . Maybe `getCallback` or `makeCallback`, `reserveCallback`

> have nodejs or similar the same concept? How do they call it?

I have no idea, but kind of doubt it as they don't wait on promises if the [main script just ends](https://github.com/nodejs/node/issues/22088) it's just not the kind of thing they need to do. 

In our case though I don't see any other way that won't be a major PITA for every user.

Reply:
My vote would be on  `registerCallback` 👍🏻

Reply:
Done including related changes  [9a12345](https://github.com/grafana/k6/pull/2228/commits/9a12345e90be182fc271a33381f3c7a76e60d51f)